### ISSUE TITLE
When using DetectDelimiter on a CSV file with a header, it should only consider delimiters in the header

### ISSUE THREAD

**Issue by thorhj** [JoshClose/CsvHelper#1803]
**Describe the bug**
If you set `CsvConfiguration.DetectDelimiter = true`, the parser will look through the entire CSV file to detect the delimiter, based on the occurences of the possible delimiters in the file. If the CSV file has a header, it is sufficient and more correct to only consider the header.

**To Reproduce**
```c#
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            var csv = @"Field1;Field2;Field3
1,2,3,4,5,6,7;1,2,3,4,5,6,7;1,2,3,4,5,6,7";
            
            using var textReader = new StringReader(csv);
            using var csvReader = new CsvReader(textReader, new CsvConfiguration(CultureInfo.InvariantCulture)
            {
                DetectDelimiter = true,
                DetectDelimiterValues = new[] { ";", "," }
            });

            var data = csvReader.GetRecords<Dto>();
        }
    }

    public class Dto
    {
        public string Field1 { get; set; }
        public string Field2 { get; set; }
        public string Field3 { get; set; }
    }
```

It fails with the following message:

```
Header with name 'Field1'[0] was not found.
Header with name 'Field2'[0] was not found.
Header with name 'Field3'[0] was not found.
If you are expecting some headers to be missing and want to ignore this validation, set the configuration HeaderValidated to null. You can also change the functionality to do something else, like logging the issue.

IReader state:
   ColumnCount: 0
   CurrentIndex: -1
   HeaderRecord:
["Field1;Field2;Field3"]
IParser state:
   ByteCount: 0
   CharCount: 22
   Row: 1
   RawRow: 1
   Count: 1
   RawRecord:
Field1;Field2;Field3
```

It detects the delimiter `,` because that character has the most occurences in the CSV file. `csvReader.HeaderRecord` has the value `string[1] { "Field1;Field2;Field3" }`.

**Expected behavior**
I would expect the parser to detect the delimiter `;` in the example above, since that's the only delimiter occuring in the header.

**Additional context**
CsvHelper 27.0.2

**Comment by thorhj** [comment]
> It's only looking at the characters in the initial buffer.

Got it, I guess in my case the header and data is just short enough to fit into the buffer then.

> I agree, but the problem is determining what the first line is. If the first line contains a \r\n but it's quoted in a field, there is no way of knowing that without parsing and it can't be parsed without a delimiter.

I hadn't even considered the possibility of newlines appearing in the header 😬

**Comment by JoshClose** [comment]
> I hadn't even considered the possibility of newlines appearing in the header 😬

That's what I'm here for!

I'll see what I can do.

**Comment by thorhj** [comment]
Thanks 👍

**Comment by JoshClose** [comment]
I changed it to check line by line. If a line doesn't contain one of the given delimiters, it will move onto the next line, until a delimiter is found or the end of the buffer. If no delimiter is found, it doesn't set one and uses whatever the config is set to.

Does that sound good?

**Comment by thorhj** [comment]
Sounds reasonable to me. 👍

**Comment by JoshClose** [comment]
The fix is in 27.0.4 on NuGet.
